pkg/sharding/handler.go

```diff
@@ -0,0 +1,35 @@
+package sharding
+
+import (
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"net/http"
+	"time"
+
+	"github.com/kcp-dev/kcp/pkg/sharding/apiserver"
+)
+
+func ServeHTTP(apiHandler http.Handler, loader *ClientLoader) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		info, ok := request.RequestInfoFrom(req.Context())
+		if !ok {
+			http.Error(w, "no request info", http.StatusInternalServerError)
+			return
+		}
+		if info == nil || !info.IsResourceRequest {
+			// not a resource request, can't handle it
+			apiHandler.ServeHTTP(w, req)
+			return
+		}
+		if sharded := req.Header.Get("X-Kubernetes-Sharded-Request"); sharded == "true" {
+			// we're being asked for our own data by some other shard, no need to fan out
+			apiHandler.ServeHTTP(w, req)
+			return
+		}
+		loader.RLock()
+		clients := loader.Clients
```

Comment:
AFAIK this is a shallow copy, so it'll potentially be subject to data races.

Reply:
Good catch, should be full copy once we expect the membership to change

Reply:
+ FIXME please 😄